### PR TITLE
Allow switching between CDF and Light Tree algorithms

### PR DESCRIPTION
**Pull request by gospodnetic**
Add option to change the light sampler algorithm under Rendering Settings -> Uniform path Tracer -> Advanced

### CODE REVIEW

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 1]
```diff
@@ -854,6 +855,16 @@ namespace
             QVBoxLayout* layout = new QVBoxLayout();
             groupbox->setLayout(layout);
 
+            QFormLayout* sublayout = create_form_layout();
+            layout->addLayout(sublayout);
+
+            QComboBox* light_sampling;
+            light_sampling = create_combobox("advanced.algorithm");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: We want to give a more descriptive name to that setting, I suggest `advanced.light_sampler.algorithm`. 

You'll also need to change the `create_direct_link()` call accordingly:

    create_direct_link("advanced.light_sampler.algorithm", "light_sampler.algorithm");

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 2]
```diff
@@ -57,11 +58,20 @@ namespace renderer
...
-BackwardLightSampler::BackwardLightSampler(const Scene& scene, const ParamArray& params)
+BackwardLightSampler::BackwardLightSampler(
+    const Scene&                        scene,
+    const ParamArray&                   params)
   : m_params(params)
   , m_emitting_triangle_hash_table(m_triangle_key_hasher)
 {
+    // Read which sampling algorithm should the sampler use.
+    const string name = params.get_required<string>("algorithm", "cdf");
+    if (name == "cdf")
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: More concisely:

    m_use_light_tree = params.get_required<string>("algorithm", "cdf") == "cdf";

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 3]
```diff
@@ -57,11 +58,20 @@ namespace renderer
 //
 // BackwardLightSampler class implementation.
 //
-
-BackwardLightSampler::BackwardLightSampler(const Scene& scene, const ParamArray& params)
+BackwardLightSampler::BackwardLightSampler(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: Insert a blank like between the comment block and the next line of code.

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 4]
```diff
@@ -82,11 +92,23 @@ BackwardLightSampler::BackwardLightSampler(const Scene& scene, const ParamArray&
         m_non_physical_lights_cdf.prepare();
     
     // Build the light tree.
-    m_light_tree_light_count = m_light_tree.build(
-        m_light_tree_lights,
-        m_emitting_triangles);
+    if (m_use_light_tree)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: Use curly braces for multiline bodies even if it's a single statement:

    if (m_use_light_tree)
    {
        m_light_tree_light_count =
            m_light_tree.build(
                m_light_tree_lights,
                m_emitting_triangles);
    }

But honestly I'd write it on a single line, it's not overly long:

    if (m_use_light_tree)
        m_light_tree_light_count = m_light_tree.build(m_light_tree_lights, m_emitting_triangles);

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 5]
```diff
@@ -82,11 +92,23 @@ BackwardLightSampler::BackwardLightSampler(const Scene& scene, const ParamArray&
...
+
+    if (!m_use_light_tree)
+    {
+        if (m_emitting_triangles_cdf.valid())
+            m_emitting_triangles_cdf.prepare();
+   
+        // Store the triangle probability densities into the emitting triangles.
+        const size_t emitting_triangle_count = m_emitting_triangles.size();
+        for (size_t i = 0; i < emitting_triangle_count; ++i)
+        m_emitting_triangles[i].m_triangle_prob = m_emitting_triangles_cdf[i].second;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: Fix indentation.

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 6]
```diff
@@ -331,10 +354,30 @@ void BackwardLightSampler::collect_emitting_triangles(
...
+                    emitting_triangle.m_triangle_prob = 0.0f;   // will be initialized once the emitting triangle CDF is built
                     emitting_triangle.m_material = material;
 
+                    const size_t emitting_triangle_index = m_emitting_triangles.size();
+                    
                     // Store the light-emitting triangle.
                     m_emitting_triangles.push_back(emitting_triangle);
+
+                    // Check if the emitting triangle will be sampled using CDF
+                    // and insert static probability into the cdf.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: More accurately:

    // When using the CDF, insert the light-emitting triangle into the CDF.

File: src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp [hunk 7]
```diff
@@ -396,6 +439,26 @@ void BackwardLightSampler::sample_non_physical_lights(
     assert(light_sample.m_probability > 0.0f);
 }
 
+void BackwardLightSampler::sample_lightset(
+    const ShadingRay::Time&             time,
+    const Vector3f&                     s,
+    const ShadingPoint&                 shading_point,
+    LightSample&                        light_sample) const
+{
+    if (m_use_light_tree)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp]: Use curly braces for multilines statements.

File: src/appleseed/renderer/kernel/lighting/lighttypes.cpp [hunk 8]
```diff
@@ -129,7 +129,7 @@ float EmittingTriangleLightSource::get_importance() const
 {
     const EDF* edf = m_triangle->m_material->get_uncached_edf();
     
-    return edf->get_uncached_max_contribution() * edf->get_uncached_importance_multiplier();
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/lighttypes.cpp]: I'm not convinced about this change.

Comment (gospodnetic) [inline src/appleseed/renderer/kernel/lighting/lighttypes.cpp]: Leftover from OSL testing, reverted

**Pull request merged**